packages/backend/src/Tabulators/IRV.ts

```diff
@@ -23,94 +42,190 @@ export function IRV(candidates: string[], votes: ballot[], nWinners = 1, randomT
+
+        let sortedVoteCounts = [...roundVoteCounts].sort((a, b) => {
+            if (a.voteCount !== b.voteCount) {
+                return (b.voteCount - a.voteCount)
             }
-            else {
-                // give vote to top rank candidate
-                roundVoteCounts[topRemainingRankIndex] += 1
+
+            // break tie in favor of candidate with higher score in previous rounds
+            for (let i = results.voteCounts.length - 1; i >= 0; i--) {
+                if (results.voteCounts[i][b.index] !== results.voteCounts[i][a.index]) {
+                    return results.voteCounts[i][b.index] - results.voteCounts[i][a.index]
+                }
             }
-        }
-        results.voteCounts.push(roundVoteCounts)
-        results.exhaustedVoteCounts.push(exhaustedVoteCount)
-        results.overVoteCounts.push(overVoteCount)
-        let voteThreshold = activeVotes.length / 2
+
+            return (summaryData.candidates[a.index].tieBreakOrder - summaryData.candidates[b.index].tieBreakOrder)
+
+        })
+
+        results.voteCounts.push(roundVoteCounts.map(c => c.voteCount.valueOf()))
+        results.exhaustedVoteCounts.push(exhaustedVotes.length)
+        results.overVoteCounts.push(exhaustedVotes.filter(ev => ev.overvote).length)
 
         // get max number of votes
         let remainingCandidatesIndexes = remainingCandidates.map(c => c.index)
-        let maxVotes = Math.max(...roundVoteCounts.filter((c, i) => remainingCandidatesIndexes.includes(i)))
-        if (maxVotes > voteThreshold || remainingCandidates.length <= 2) {
+
+        let maxVotes = sortedVoteCounts[0].voteCount
+        let nActiveVotes = candidateVotes.map(c => c.length).reduce((a, b) => a + b, 0)
+        if (!proportional) {
+            quota = Math.floor(nActiveVotes /2 + 1)
+        }
+
+        if (maxVotes >= quota) {
             // candidate meets the threshold
             // get index of winning candidate
-            let winningCandidateIndex = roundVoteCounts.indexOf(maxVotes)
+            let winningCandidateIndex = sortedVoteCounts[0].index
             // add winner, remove from remaining candidates
             results.elected.push(summaryData.candidates[winningCandidateIndex])
-            remainingCandidates = remainingCandidates.filter(c => !results.elected.includes(c))
-            // TODO: Some STV stuff
+            roundResults.winners.push(summaryData.candidates[winningCandidateIndex])
+            if (proportional) {
+                remainingCandidates = remainingCandidates.filter(c => !results.elected.includes(c))
+                let fractionalSurplus = new Fraction(maxVotes - quota).div(maxVotes)
+                let winningCandidateVotes = candidateVotes[winningCandidateIndex]
+                candidateVotes[winningCandidateIndex] = []
+                winningCandidateVotes.forEach(vote => {
+                    vote.weight = vote.weight.mul(fractionalSurplus).floor(5)
```

Comment:
The weight is already a fraction type, why do we need to floor it?

Reply:
This is also used in Scottish STV. We don't really need it but wanted to be consistent. Even with floor we still need fractions because of floating point errors (0.2 + 0.1 < 0.3). What the fraction floor method does is set the denominator to 10000 so it keeps the fraction sizes manageable and consistent, rather than 12341245/12342343547 you have 999/10000.